src/Compilers/Core/Portable/SourceGeneration/GeneratorState.cs

```diff
@@ -2,31 +2,103 @@
         {
-            this.Sources = sources;
+            this.SourceTexts = sourceTexts;
+            this.Trees = trees;
             this.Info = info;
+            this.Diagnostics = diagnostics;
+            this.SyntaxReceiver = syntaxReceiver;
+            this.Exception = exception;
         }
 
-        internal ImmutableDictionary<GeneratedSourceText, SyntaxTree> Sources { get; }
+        internal ImmutableArray<GeneratedSourceText> SourceTexts { get; }
+
+        internal ImmutableArray<SyntaxTree> Trees { get; }
 
         internal GeneratorInfo Info { get; }
 
-        internal GeneratorState WithSources(ImmutableDictionary<GeneratedSourceText, SyntaxTree> sources)
+        internal ISyntaxReceiver? SyntaxReceiver { get; }
+
+        internal Exception? Exception { get; }
+
+        internal ImmutableArray<Diagnostic> Diagnostics { get; }
+
+        /// <summary>
+        /// Called to save the created syntax receiver
+        /// </summary>
+        /// <remarks>
+        /// We retain the existing state, as it may be needed for the upcoming generation pass.
+        /// </remarks>
+        internal GeneratorState WithReceiver(ISyntaxReceiver syntaxReceiver)
+        {
+            Debug.Assert(this.SyntaxReceiver is null);
+            return new GeneratorState(this.Info,
+                                      sourceTexts: this.SourceTexts,
+                                      trees: this.Trees,
+                                      diagnostics: this.Diagnostics,
+                                      syntaxReceiver: syntaxReceiver,
+                                      exception: null);
+        }
+
+        /// <summary>
+        /// Called when the generator has a result to store.
+        /// </summary>
+        /// <remarks>
+        /// We discard any receiver that was set as it is no longer needed.
+        /// We discard any saved exception as it must refer to a previous iteration.
+        /// </remarks>
+        internal GeneratorState WithResult(ImmutableArray<GeneratedSourceText> sourceTexts,
+                                          ImmutableArray<SyntaxTree> trees,
+                                          ImmutableArray<Diagnostic> diagnostics)
+        {
+            Debug.Assert(sourceTexts.Length == trees.Length);
+            return new GeneratorState(this.Info,
+                                      sourceTexts,
+                                      trees,
+                                      diagnostics,
+                                      syntaxReceiver: null,
+                                      exception: null);
+        }
+
+        /// <summary>
+        /// Called when the generator threw an exception
+        /// </summary>
+        /// <remarks>
+        /// We discard any other saved state, as it must refer to a previous iteration and is no longer needed.
+        /// </remarks>
+        internal GeneratorState WithError(Exception e, Diagnostic diagnostic)
```

Comment:
>WithError [](start = 32, length = 9)

I know Jared said to use `With` as a prefix, but I find this confusing as most of these methods do not actually make copies, they move us from state to state, dropping data as appropriate. Perhaps `MoveTo` is a more correct prefix? Basically, I'd never expect a wither to drop things.

Reply:
Right. This is too confusing, so i've decided to remove all the 'state-machine' verbiage and not-with withers. Instead there are just constructors with different param sets. 

I think it acheives the exact same thing but makes it *far* clearer what it's doing.

---
In reply to: [466072205](https://github.com/dotnet/roslyn/pull/46296#discussion_r466072205) [](ancestors = 466072205)